### PR TITLE
Switch core master base images (kube-apiserver, kube-scheduler) from debian to distroless

### PR DESCRIPTION
**Pull request by dims**
Kudos to @yuwenma for all the earlier work on this that made this PR possible!

Premise : Since the PR for switching on `--log-file` in the manifests ran into trouble and got reverted a few times, trying another approach here. The key idea here is that we need to
- Keep the same behavior we have now with redirection of stdout/stderr `1>>/var/log/kube-apiserver.log 2>&1`
- While avoiding a dependency on bash or any other shell

So we basically need a go based runner which redirects stdout/stderr. See [go-runner.go](https://github.com/dims/go-runner/blob/master/go-runner.go). Then we need to wrap this go-runner in a [distroless image](https://github.com/dims/go-runner/blob/master/Dockerfile) which we can then use in both the bazel and make based builds. So that's what we ended up in this PR.

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Please see the image repo https://github.com/dims/go-runner, happy to add that repo to k/k or elsewhere once we validate this approach works

**Which issue(s) this PR fixes**:

Related to KEP https://github.com/kubernetes/enhancements/tree/master/keps/sig-release/1729-rebase-images-to-distroless

**Special notes for your reviewer**:
Borrowed code from previous experiments by @yuwenma from https://github.com/kubernetes/kubernetes/pull/75306 and https://github.com/kubernetes/kubernetes/pull/83390

**Does this PR introduce a user-facing change?**:

```release-note
Switch core master base images (kube-apiserver, kube-scheduler) from debian to distroless
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by liggitt** [comment]
Does the go-runner behave the same as the bash redirection with interleaving stdout/stderr lines coherently?

I vaguely remember running into related issues with the way go launches sub processes for tests and combines the stdout/stderr output (https://github.com/golang/go/issues/33419#issuecomment-517800582)

If bash redirection already interleaves lines incoherently, I guess this is no worse. If it interleaves line-by-line, then we need to figure out a way to do the same here.

**Comment by dims** [comment]
@liggitt bash already interleaves. Just tried this test https://gist.github.com/dims/da574174817135b5ddf100b9ee89e8e6

**Comment by mm4tt** [comment]
Nice, I'll see what we can do to scale test it

**Comment by mm4tt** [comment]
+ @tosi3k, who will help with running the tests

**Comment by dims** [comment]
@mm4tt @tosi3k awesome! thanks.

**Comment by tosi3k** [comment]
@dims just to confirm before tests execution - is c80b2c6 the commit we want the tests to run on top of?

If so, we will run the tests tomorrow and come back here with conclusions on the day after tomorrow (on Wednesday, to be precise).

**Comment by dims** [comment]
@tosi3k confirmed. yes please!

**Comment by dekkagaijin** [comment]
If this gives good enough performance at scale, LGTM. We'd want to move the `go-runner` base image to a more secure repo, but it does qualify as `distroless` as far as security is concerned

**Comment by tallclair** [comment]
I haven't had a chance to carefully read through go-runner, but can someone (@dims ?) give me a quick explanation of why go-runner solves the performance problems, and the solution can't just be integrated back into klog? In other words, why does writing to the log files in a separate process solve the problems?

**Comment by liggitt** [comment]
we are not guaranteed all stderr/stdout output is generated via klog

**Comment by tallclair** [comment]
@liggitt klog could "seize control" of stdout/err by replacing the global stdout/stderr variable with a pipe.

**Comment by dims** [comment]
@tallclair the scalability team is running a test with this PR, let's see if it actually works. go-runner just mimics what we used to do with bash redirection as closely as possible in the hope that it works at this point until we have confirmation that it does.

**Comment by dims** [comment]
> If this gives good enough performance at scale, LGTM. We'd want to move the `go-runner` base image to a more secure repo, but it does qualify as `distroless` as far as security is concerned

Yes of course. +1 @dekkagaijin

**Comment by immutableT** [comment]
> Keep the same behavior we have now with redirection of stdout/stderr 1>>/var/log/kube-apiserver.log 2>&1

Why is this required/desirable?

**Comment by dims** [comment]
@immutableT I am assuming capturing kube-apiserver logs is desirable :)

ideally we want to use klog's `log-file` to capture the api server logs (see [here](https://github.com/kubernetes/kubernetes/pull/78466/files#diff-6e3b476b9225d1213dc6ad13e453fc16R1863-R1865)) but that ran into scalability issues with log rotation, but that attempt got reverted in https://github.com/kubernetes/kubernetes/pull/83390 (see discussion there). Also see https://github.com/kubernetes/kubernetes/pull/75306 which we would have needed after 78466 landed.

So this PR is an attempt to get us to distroless images by picking up pieces from the attempts above but using go-runner instead of bash.

**Comment by immutableT** [comment]
Yes, I agree :)
I was trying to ask whether we considered/tested the use of kubelet's log rotation feature (#59898) since the root cause of failures of #83390 seems to be related to that?

**Comment by dims** [comment]
@immutableT that feature does not work for the majority of the CI jobs that use docker shim (it's specific to non-docker CRI based impls), also there's a serious problem with files cleanup that is still being worked on in https://github.com/kubernetes/kubernetes/issues/90334

**Comment by tosi3k** [comment]
@dims the experimental 5k test run passed. Here are Grafana dashboards from a [baseline run](http://35.226.146.156:3000/d/uzJNy03Wk/master-dashboard?orgId=1&var-source=distroless-images-baseline&var-etcd_type=All&var-etcd_operation=All&var-verb=All&var-resource=All) and the [experimental run](http://35.226.146.156:3000/d/uzJNy03Wk/master-dashboard?orgId=1&var-source=distroless-images-experiment&var-etcd_type=All&var-etcd_operation=All&var-verb=All&var-resource=All).

The results look good - I don't see any distinct differences between the runs but that's just from my novice's perspective :).

**Comment by mm4tt** [comment]
Thanks, @tosi3k!
I double checked and I can confirm what @tosi3k wrote, I don't see any significant performance impact of this PR. Looks like this approach will work

**Comment by dims** [comment]
@tosi3k @mm4tt Excellent News!!! wow. will add another PR for `go-runner`, get that into k8s.gcr.io and then revise this PR to use the image

**Comment by justaugustus** [comment]
Explicit hold on this, until I review (as it affects the releases). Please do not remove it.
/hold

I should be able to get to this later today. Worst case tomorrow.

File: build/common.sh
Comment (justaugustus) [inline build/common.sh:102]:
```suggestion
    "kube-apiserver,${KUBE_BASE_IMAGE_REGISTRY}/go-runner:v0.1.1"
```

File: build/common.sh
Comment (justaugustus) [inline build/common.sh:103]:
```suggestion
    "kube-controller-manager,${KUBE_BASE_IMAGE_REGISTRY}/go-runner:v0.1.1"
```

File: build/common.sh
Comment (justaugustus) [inline build/common.sh:104]:
```suggestion
    "kube-scheduler,${KUBE_BASE_IMAGE_REGISTRY}/go-runner:v0.1.1"
```

File: build/workspace.bzl
Comment (justaugustus) [inline build/workspace.bzl:101]:
```suggestion

# Use skopeo to find these values: https://github.com/containers/skopeo
#
# Example
# Manifest: skopeo inspect docker://gcr.io/k8s-staging-build-image/go-runner:v0.1.1
# Arches: skopeo inspect --raw docker://gcr.io/k8s-staging-build-image/go-runner:v0.1.1
```

File: build/workspace.bzl
Comment (justaugustus) [inline build/workspace.bzl:103]: Needs the v0.1.1 digests once promoted.
Same w/ the arch digests :point_down:

File: build/workspace.bzl
Comment (justaugustus) [inline build/workspace.bzl:125]:
```suggestion
            tag = "v0.1.1",  # ignored, but kept here for documentation
```

File: build/workspace.bzl
Comment (justaugustus) [inline build/workspace.bzl:123]:
```suggestion
            registry = "us.gcr.io/k8s-artifacts-prod/build-image",
```

**Review by justaugustus — CHANGES_REQUESTED** [review]
@dims -- Made some code suggestions.

You'll also need to make entries in `build/dependencies.yaml`: https://github.com/kubernetes/kubernetes/blob/90f7ae54acbbab22460fe5fabded02d9b912b834/build/dependencies.yaml#L78-L104

Also, can we pull this into its own Dockerfile? :point_down: 
https://github.com/kubernetes/kubernetes/blob/90f7ae54acbbab22460fe5fabded02d9b912b834/build/lib/release.sh#L378-L379

...and make sure the :point_up: `${base_image}` references the go-runner image.

File: build/common.sh
Comment (dims) [reply build/common.sh:102]: fixing

**Comment by justaugustus** [comment]
@dims -- This is looking really great! I'll run a mock stage/release once this merges as well to make sure we're in good shape.

/lgtm
/approve
/assign @BenTheElder @cblecker @fejta

**Review by fejta — COMMENTED** [review]
/approve

**Comment by justaugustus** [comment]
(the suspense is killing me!)

**Comment by BenTheElder** [comment]
/retest
flexvolume gce flake

**Comment by BenTheElder** [comment]
/lgtm
/approve
quite exciting.

**Comment by justaugustus** [comment]
`[sig-storage] Flexvolumes should be mountable when attachable`
/retest

Hmmm... https://github.com/kubernetes/kubernetes/issues/78737
FYI @kubernetes/sig-storage-bugs

**Comment by BenTheElder** [comment]
Uh `pull-kubernetes-e2e-gce` has not passed even once on the current commit.
In fact, it has failed 10 times in a row.
/hold

**Comment by BenTheElder** [comment]
I remember this now. The controller manager runs flexvolume plugins, and flexvolume plugins include our e2e plugins may be shell scripts. You can't rip out the userspace from controller manager without breaking flexvolume.
/lgtm cancel

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-storage/flexvolume.md

init, attach, detach, wait for attach, checking if attached are all called from the controller manager, which is done via an exec model. While the binary called comes from a path on the host, it is invoked in the controller-manager container filesystem (or kubelet).

Currently they can reasonably expect a generic linux userspace ... debian specifically for controller manager.

File: build/BUILD
Comment (justaugustus) [inline build/BUILD:38]:
```suggestion
        "base": "@debian-base-{ARCH}//image",
```

File: build/common.sh
Comment (justaugustus) [inline build/common.sh:104]:
```suggestion
    "kube-controller-manager,${KUBE_BASE_IMAGE_REGISTRY}/debian-base-${arch}:${debian_base_version}"
```

File: cluster/gce/manifests/kube-controller-manager.manifest
Comment (justaugustus) [inline cluster/gce/manifests/kube-controller-manager.manifest:29]: Revert this change.

**Review by justaugustus — CHANGES_REQUESTED** [review]
@BenTheElder -- would I be correct in the assumption that removing the controller-manager changes should get us back in a good state?

If so, @dims, I've made a few code suggestions on what to revert.

EDIT: @dims -- In places where we can't/shouldn't use `go-runner` e.g., `kube-controller-manager`, `kube-proxy`, can you add a note referencing the [base image exception list](https://git.k8s.io/sig-release/release-engineering/baseimage-exception-list.md)?

**Comment by BenTheElder** [comment]
As far as I know controller-manager / flexvolume is the only thing like this, reverting it should be sufficient but unfortunate, we'll still be stuck shipping these images ...

I had some conversations with @yuwenma and the GKE node team about this in the past when we were trying to figure out @yuwenma's attempt at this.

IIRC per @msau42 the idea was that eventually flexvolume could be deprecated out, but probably not until CSI was further along, or something along those lines ...

**Comment by justaugustus** [comment]
> As far as I know controller-manager / flexvolume is the only thing like this, reverting it should be sufficient but unfortunate, we'll still be stuck shipping these images ...

Okay, that was what I was getting from skimming through some of the docs/issues.
FWIW, we'll be still be required to ship `debian-base` (to support `debian-iptables` and others on the [exceptions list](https://github.com/kubernetes/sig-release/blob/master/release-engineering/baseimage-exception-list.md)).

**Comment by dims** [comment]
/hold cancel

@justaugustus @BenTheElder removed controller-manager from this PR. Let's deal with that later.

**Comment by justaugustus** [comment]
> EDIT: @dims -- In places where we can't/shouldn't use `go-runner` e.g., `kube-controller-manager`, `kube-proxy`, can you add a note referencing the [base image exception list](https://git.k8s.io/sig-release/release-engineering/baseimage-exception-list.md)?

@dims -- Can you address this as well? I want to make sure we deter changes to the images and provide context about why.

**Comment by dims** [comment]
@justaugustus not yet. i want this to merge and i am looking at options for controller-manager, looking at logs now.

**Pull request merged**
